src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1413,86 +1389,68 @@ final ForkJoinTask<?> peek() {
+                    int cap, b, k, nb; ForkJoinTask<?>[] a;
+                    if ((a = array) == null || (cap = a.length) <= 0)
+                        break;
+                    long kp = slotOffset(k = (cap - 1) & (b = base));
+                    int nk = (nb = b + 1) & (cap - 1); // next slot
+                    int sk = (b + 2) & (cap - 1);      // 2nd slot ahead
+                    ForkJoinTask<?> t = a[k];
+                    U.loadFence();
+                    if (base != b)
+                        ;                              // inconsistent
+                    else if (t == null) {
+                        if (a[sk] == null && a[nk] == null && a[k] == null &&
+                            top - b <= 0)
+                            break;                     // empty
+                        if (stalled)
+                            Thread.onSpinWait();
+                        stalled = true;
+                        U.loadFence();                 // reread
+                    }
+                    else if (U.compareAndSetReference(a, kp, t, null)) {
                         updateBase(nb);
-                        if (a[nk] != null && pool != null)
-                            pool.signalWork(a, nk); // propagate
                         return t;
                     }
-                    if (o == null && a[nk] == null && array == a &&
-                        (phase & (IDLE | 1)) != 0 && top - base <= 0)
-                        break;                    // empty
                 }
             }
             return null;
         }
 
+        // specialized execution methods
+
         /**
-         * Tries to poll next task in FIFO order, failing without
-         * retries on contention or stalls. Used only by topLevelExec
-         * to repoll from the queue obtained from pool.scan.
+         * Runs the given task, as well as remaining local tasks, plus
+         * those from src queue that can be taken without interference.
          */
-        private ForkJoinTask<?> tryPoll() {
-            ForkJoinTask<?>[] a; int cap;
-            if ((a = array) != null && (cap = a.length) > 0) {
-                for (int b = base, k;;) {  // loop only if inconsistent
-                    ForkJoinTask<?> t = a[k = b & (cap - 1)];
-                    U.loadFence();
-                    if (b == (b = base)) {
-                        Object o;
-                        if (t == null)
-                            o = a[k];
-                        else if (t == (o = U.compareAndExchangeReference(
-                                           a, slotOffset(k), t, null))) {
-                            updateBase(b + 1);
-                            return t;
-                        }
-                        if (o == null)
+        final void topLevelExec(ForkJoinTask<?> task, WorkQueue src,
+                                int srcBase, int cfg) {
+            if (task != null && src != null) {
```

Comment:
@DougLea Is this ever invoked with a null-task and a null-queue? 🤔

Reply:
No. This is one of many hoisted null checks that help compilers avoid throwing null pointer exceptions where we don't want them.  Usually I mark these with " // always true", which I'll add here.